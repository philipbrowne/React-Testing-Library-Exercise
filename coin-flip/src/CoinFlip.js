import React, { useState } from 'react';
import heads from './dime-front.png';
import tails from './dime-back.png';
import Coin from './Coin';

const CoinFlip = () => {
  const [coinSide, setCoinSide] = useState('');
  const [flipCount, setFlipCount] = useState(0);
  const [headCount, setHeadCount] = useState(0);
  const [tailCount, setTailCount] = useState(0);
  const [isCoinFlipped, setIsCoinFlipped] = useState(false);
  const flipCoin = () => {
    setFlipCount(flipCount + 1);
    setIsCoinFlipped(true);
    const rand = Math.random();
    if (rand >= 0.5) {
      setCoinSide(heads);
      checkSide();
    } else {
      setCoinSide(tails);
      checkSide();
    }
  };
  const checkSide = () => {
    if (coinSide === heads) {
      setHeadCount(headCount + 1);
    } else if (coinSide === tails) {
      setTailCount(tailCount + 1);
    }
  };
  return (
    <div className="CoinFlip">
      <h1 className="CoinFlip-header">Let's flip a coin!</h1>
      <button
        onClick={() => {
          flipCoin();
        }}
        data-testid="button"
      >
        FLIP ME!
      </button>
      {isCoinFlipped && <Coin img={coinSide} data-testid="coin" />}
      <p>
        Out of {flipCount} flips there have been {headCount} heads and{' '}
        {tailCount} tails.
      </p>
    </div>
  );
};

export default CoinFlip;
