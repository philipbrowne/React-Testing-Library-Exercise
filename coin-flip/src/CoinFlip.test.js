import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import CoinFlip from './CoinFlip';

beforeEach(function () {
  jest
    .spyOn(Math, 'random')
    .mockReturnValueOnce(0.25)
    .mockReturnValueOnce(0.75);
});

afterEach(function () {
  Math.random.mockRestore();
});

it('renders without crashing', () => {
  render(<CoinFlip />);
});

it('should match snapshot', () => {
  const { asFragment } = render(<CoinFlip />);
  expect(asFragment()).toMatchSnapshot();
});

it('does not show coin on load, but shows after click of button', () => {
  const { queryByTestId, queryByAltText } = render(<CoinFlip />);
  const coin = queryByTestId('coin');
  expect(coin).not.toBeInTheDocument();
  const btn = queryByTestId('button');
  fireEvent.click(btn);
  const coinImg = queryByAltText('coin');
  expect(coinImg).toBeInTheDocument();
});

it('shows tails on the first flip and heads on the second flip', () => {
  const { queryByAltText, queryByTestId } = render(<CoinFlip />);
  const btn = queryByTestId('button');
  fireEvent.click(btn);
  const coinImg = queryByAltText('coin');
  expect(coinImg.src).toContain('dime-back.png');
  fireEvent.click(btn);
  expect(coinImg.src).toContain('dime-front.png');
});
