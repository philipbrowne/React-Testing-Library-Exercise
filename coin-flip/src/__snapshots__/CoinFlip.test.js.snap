// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="CoinFlip"
  >
    <h1
      class="CoinFlip-header"
    >
      Let's flip a coin!
    </h1>
    <button
      data-testid="button"
    >
      FLIP ME!
    </button>
    <p>
      Out of 0 flips there have been 0 heads and 0 tails.
    </p>
  </div>
</DocumentFragment>
`;
